### PR TITLE
Made Website into Progressive Web App

### PR DESCRIPTION
**Pull request by AdvaitDhingra**
Progressive Web Apps give a native app like experience for websites. When root.cern is opened one can install it and it will run like a native application. 

To achieve this I added a manifest.json file and made a service worker in Javascript.

It passed the Google Lighthouse test:

Screenshot from 2021-07-14 22-32-50

And it is installable on both iOS and Android. 

When one opens root.cern one gets a prompt to install it.

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
I suppose that's mostly useful for the reference guide? I know the forum is "installable" already. Can you describe the prupose of this, @AdvaitDhingra ?

**Comment by AdvaitDhingra** [comment]
> I suppose that's mostly useful for the reference guide? I know the forum is "installable" already. Can you describe the prupose of this, @AdvaitDhingra ?

Yes, the forum is already installable. This update would allow the root website to be used (to a certain extent) offline and also installable as a pwa so that you wouldn't always have to open your browser and type roo.cern.

**Comment by AdvaitDhingra** [comment]
Since the reference guide isn't a standalone website I made the whole root.cern a pwa

**Comment by Axel-Naumann** [comment]
But the failures in the HTML proofer tell me that you needed to modify each web page (by changing the header). That would likely mean that the reference guild won't be affected by this PR? (That's not an argument against merging this - it's maybe a request to also add this to [ROOT's doxygen header](https://github.com/root-project/root/blob/master/documentation/doxygen/htmlheader.html) :-)

**Comment by AdvaitDhingra** [comment]
> But the failures in the HTML proofer tell me that you needed to modify each web page (by changing the header). That would likely mean that the reference guild won't be affected by this PR? (That's not an argument against merging this - it's maybe a request to also add this to [ROOT's doxygen header](https://github.com/root-project/root/blob/master/documentation/doxygen/htmlheader.html) :-)

Oh, I didn't know there were more headers 😅😂. I'll add this to the others too 🙂

File: serviceWorker.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+self.addEventListener("install", function (event) {
+  console.log("Install Event processing");
+
+  event.waitUntil(
+    caches.open(CACHE).then(function (cache) {
+      console.log("Cached offline page during install");
+
+      if (offlineFallbackPage === "ToDo-replace-this-name.html") {
+        return cache.add(new Response("Update the value of the offlineFallbackPage constant in the serviceworker."));
+      }
```
Comment (jalopezg-r00t) [inline serviceWorker.js:15]: Do we really need this?  ;-)

File: serviceWorker.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+// If any fetch fails, it will look for the request in the cache and serve it from there first
+self.addEventListener("fetch", function (event) {
+  if (event.request.method !== "GET") return;
+
+  event.respondWith(
+    fetch(event.request)
+      .then(function (response) {
+        console.log("Add page to offline cache: " + response.url);
+
```
Comment (jalopezg-r00t) [inline serviceWorker.js:29]: I think we could also remove this.

File: manifest.json [hunk 3]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "ROOT: analyzing petabytes of data, scientifically.",
+  "short_name": "ROOT",
+  "description": "ROOT Website",
+  "theme_color": "#346295 ",
+  "background_color": "#346295 ",
```
Comment (jalopezg-r00t) [inline manifest.json:6]: Maybe remove the extra spaces at the end.

**Review by jalopezg-r00t — CHANGES_REQUESTED** [review]
Thanks for the contribution!  Please, consider also the following suggestions.  :-)

File: serviceWorker.js (hunk 1 above)
Comment (AdvaitDhingra) [reply serviceWorker.js:15]: Now that I think about it, not really 😅

File: serviceWorker.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
+const CACHE = "root-offline";
```
Comment (Axel-Naumann) [inline serviceWorker.js:1]: Where does this file come from? Did you write this yourself? If not, what is the license of the file and where does it come from?

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
Clarify origin

File: serviceWorker.js (hunk 4 above)
Comment (AdvaitDhingra) [reply serviceWorker.js:1]: No it's just a name I came up with of the cache file.

**Comment by AdvaitDhingra** [comment]
> Clarify origin

What exactly do you mean by that, sorry 😅 . Should I clarfiy the directory where manifest.json can be found?

**Comment by Axel-Naumann** [comment]
@AdvaitDhingra sorry for the confusion. I would like to know where the files you are proposing here are coming from: whether you have authored them yourself, whether you adapted them, or whether they are straight copies, for each of the files, and in case where you did not write them completely yourself, what their origin and license is. That's needed to determine whether we can include this. Referencing the origin is also simply good style :-)

**Comment by AdvaitDhingra** [comment]
> @AdvaitDhingra sorry for the confusion. I would like to know where the files you are proposing here are coming from: whether you have authored them yourself, whether you adapted them, or whether they are straight copies, for each of the files, and in case where you did not write them completely yourself, what their origin and license is. That's needed to determine whether we can include this. Referencing the origin is also simply good style :-)

Ah ok. I used online guides but the code used is available by many sources. I can try to find an origin.

**Comment by Axel-Naumann** [comment]
That's fine - if there is no single source that you can point to where you copied a file from, but you have "cooked" this out of multiple online sources, then we don't have a copyright / license question to address.

**Comment by AdvaitDhingra** [comment]
> That's fine - if there is no single source that you can point to where you copied a file from, but you have "cooked" this out of multiple online sources, then we don't have a copyright / license question to address.

Okay, thanks. In the mean time I'll look for a way to link the manifest correctly

**Comment by Axel-Naumann** [comment]
See e.g. https://github.com/root-project/web/blob/7389c44d030bee71d90bfc7f45f5ea2de011a469/_includes/footer.html#L4 for how to create such a link

**Comment by AdvaitDhingra** [comment]
> See e.g.
> 
> https://github.com/root-project/web/blob/7389c44d030bee71d90bfc7f45f5ea2de011a469/_includes/footer.html#L4
> 
> for how to create such a link

Okay, thanks

**Comment by AdvaitDhingra** [comment]
The problem is that the index.html file is in the root directoy, so for that manifest.json is just manifest.json but for the others (refrence for example) its ../manifest.json .

**Comment by AdvaitDhingra** [comment]
That did it. Instead of {{site.baseurl}} I used {{site.url}} . Since baseurl was static and url is just root.cern

**Comment by AdvaitDhingra** [comment]
@Axel-Naumann is everything good?

**Comment by Axel-Naumann** [comment]
Possibly; I'd like @couet to have a look (for instance at the relative links) once he's back next week.

**Review by couet — APPROVED** [review]
It looks good to me. All tests passed. Let see what extra functionalities this will bring.

**Comment by AdvaitDhingra** [comment]
Thanks @couet . Can I squash and merge it?

**Pull request closed without merging**